### PR TITLE
[basic-auth] Better error in secret validation

### PR DESCRIPTION
**Pull request by shvgn**
## Description

Enhanced the error message when there are unexpected number of fields in credentials secret.

## Why do we need it, and what problem does it solve?

Better describes user error in Deckhouse logs

## What is the expected result?

The error should be more consistent with what the secret state

## Checklist
- [ ] The code is covered by unit tests.
- [ ] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [ ] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: basic-auth
type: chore
summary: Improved error message on unexpected number of fields in the credentials secret
```

### CODE REVIEW

File: ee/fe/modules/500-basic-auth/hooks/generate_passwords.go [hunk 1]
```diff
@@ -117,9 +117,13 @@ func restorePasswordFromSnapshot(snapshot []go_hook.FilterResult) (string, error
...
 	}
+
 	// Only one field is expected.
-	if len(secretData) != 1 {
-		return "", fmt.Errorf("secret/%s has many fields", secretName)
+	if len(secretData) == 0 {
+		return "", fmt.Errorf("secret/%s has no fields", secretName)
+	}
+	if len(secretData) > 1 {
+		return "", fmt.Errorf("secret/%s has multiple fields", secretName)
```
Comment (shvgn) [inline ee/fe/modules/500-basic-auth/hooks/generate_passwords.go:126]:
```suggestion
		return "", fmt.Errorf("secret/%s has multiple fields, only 'htpasswd' field is expected", secretName)
```

File: ee/fe/modules/500-basic-auth/hooks/generate_passwords.go [hunk 2]
```diff
@@ -117,9 +117,13 @@ func restorePasswordFromSnapshot(snapshot []go_hook.FilterResult) (string, error
 	if !ok {
 		return "", fmt.Errorf("secret/%s has empty data", secretName)
 	}
+
 	// Only one field is expected.
-	if len(secretData) != 1 {
-		return "", fmt.Errorf("secret/%s has many fields", secretName)
+	if len(secretData) == 0 {
+		return "", fmt.Errorf("secret/%s has no fields", secretName)
```
Comment (shvgn) [inline ee/fe/modules/500-basic-auth/hooks/generate_passwords.go:123]:
```suggestion
		return "", fmt.Errorf("'htpasswd' field is missing in secret/%s", secretName)
```

File: ee/fe/modules/500-basic-auth/hooks/generate_passwords.go (hunk 1 above)
Comment (shvgn) [reply ee/fe/modules/500-basic-auth/hooks/generate_passwords.go:126]: Following current logic, why do we care about extra fields?

Comment (shvgn) [reply ee/fe/modules/500-basic-auth/hooks/generate_passwords.go]: Did it!

Comment (shvgn) [reply ee/fe/modules/500-basic-auth/hooks/generate_passwords.go]: But it broke tests, so I reverted it back

**Review by diafour — APPROVED** [review]
lgtm

**Review by diafour — APPROVED** [review]
lgtm

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @diafour. I've started the workflow for label 'e2e/run/gcp'...

**Pull request merged**